lookout/style/format/code_generator.py

```diff
@@ -59,7 +61,7 @@ def apply_predicted_y(self, vnodes: Sequence[VirtualNode],
                           vnodes_y: Sequence[VirtualNode],
                           rule_winners: Sequence[int],
                           rules: Rules,
-                          ) -> List[VirtualNode]:
+                          ) -> Sequence[VirtualNode]:
```

Comment:
I don't think it is actually good practice after having coded with types for a few months. The reason is that you can't actually chain functions this way without copying the input and have it typecheck. This means that we should just code to interfaces as usual in any language (so in input and output types we should decide on the most appropriate type). To me it should be `Sequence` (not `MutableSequence`) for maintenance sake but that is in itself debatable.

Reply:
Yes, if we return a list, then our contract must state that we return a list.

In general, the difference is:

* If you annotate a function argument, you need to be as broad as possible
* If you annotate a return value, you need to be as precise as possible

Reply:
how do you address the situation above then?

Reply:
I am confused, why I need to copy something?

Reply:
Just when you pass around some value as we do in many functions in format analyzer. Since it gets the type of the argument in the function you need to remake it of the specific type to comply to the return type. It's a problem we have in many places where we have lists everywhere but if we want to program to interfaces (sequences) we'd have to remake them in plenty of functions or cast them.

Reply:
Type annotations have nothing to do with how Python code works. I always thought that `Sequence` is `Iterable` + `__len__()`. If it implies immutability, then we must change many existing signatures because we have lists everywhere I annotated. Actually, the very fact that I got confused with these interfaces is a strong reason to abandon interfaces completely TBH, because I rarely confuse simple things.

If a function returns a list, we must annotate the return type as `List`. Otherwise, if it returns a tuple, it must be annotated as `Tuple`.

Reply:
We don't want general functions, because we code in Python, not Scala or C++ 😄 

Actually, Sequence does not guarantee that the object is read-only, from what I see in https://docs.python.org/3/glossary.html#term-sequence It does not guarantee that the object is mutable, and that's quite different.

Reply:
So my annotations are correct after all.

Reply:
yes, those hints are like typeclasses in haskell, they only define what's present, not what's absent

Reply:
actually there is a way to encode that precisely and limit our problems which would be

```python
T_Sequence = TypeVar("T_Sequence", bound=Sequence)
def f(a: T_Sequence) -> T_Sequence:
    return a
```
and now the type is kept as is even through the function and it is still the most general possible.

Reply:
given that it's fine to keep the other guideline of broadest for args and narrowest for results :)

Reply:
(returned back from the Spanish class)

So I get the point of copying, these are my considerations:

- If we know that the function always returns a list (as in this snippet, `apply_predicted_y` does return a list), we should annotate with `List`. the same with tuples, etc.
- If we pass generic function arguments forward, we should definitely save the original, generic annotation, as was noted in https://github.com/src-d/style-analyzer/pull/505#discussion_r243333005
- It is ok to annotate a function argument with `Sequence`, but it's not OK to mutate it because `Sequence` does not guarantee mutability. So this effectively behaves as "immutable" marker.
- TypeVar-s go far along the typing road, and IMHO start raping the language. Let's use them as an experiment in those cases where you think we really need them and see if we hit any problems.

Reply:
2 thoughts:

- always following the narrowest type seems limiting to me. For example: annotating https://github.com/src-d/style-analyzer/blob/master/lookout/style/format/feature_extractor.py#L195 with `List` only (and not `Sequence`) doesn't give us the nice type hint that we are messing with state that is not meant to be messed with. We could use the possibility of using more general types to our advantage.
- the only way to pass generic annotation forward in a correct way is through TypeVar. I don't consider them an advanced type feature since they are really necessary to make any correct Java program, for example. They are indeed ugly though currently since they need an additional external declaration :(

@zurk you can review the rest btw since this PR was mainly to assert my understanding of your code since I have to edit it.

Reply:
@m09 I don't get both points.

* `List` is a `Sequence`. The code returns a list. I don't see any hints. However, I see the problem with the generic annotation because the code should not be generic as a rule of a thumb and the programmers should be confident in their exact types. Otherwise, we get a combinatorial explosion of type possibilities, and since nobody is perfect, Python is not Haskell and we end up with a huge mess.
* I do not see any problems with passing outside a `Sequence[int]` if you get a `Sequence[int]`. TypeVar-s are advanced indeed because `0.01%` of Python developers really know how they work before reading the docs. Myself included. I am really concerned with code simplicity, and the benefits of TypeVar-s do not overweigh the pain in the ass they do to people who like dynamic types. Again, Python is not Haskell, and what works for Haskell is very unnatural for Python.

I have already had this many times in C++. Somebody suggests to use a cool language feature which is "the only way" and "the real solution", then the same people revisit their code in 6 months and spend days do finally understand how to refactor without breaking the whole world. Same here: a small refactoring happens and all those typing hints start messing in the way, painfully. Besides. some people are going to make errors in updates because it is hard to update complex hints, etc.

Now you can say that if we use mypy then everything gets checked and not painful. However, mypy is really raw, a negligible fraction of the community uses it and is against the Python's philosophy, after all. We used to code without type hints for 20 years, Django was written without type hints, and it is HUGE.

We should type hints as a lightweight way to guide the engineers who to use our code, nothing more. We should not go crazy with describing everything to high detail, and should not spend much time on typing things. Instead, we should spend this time on writing more tests.